### PR TITLE
Feature/simple local cache

### PR DESCRIPTION
**Pull request by theo-damiani**
SimpleLocalCache: addressed #235
Refactor viewModel: addressed #196 
Fix Bug: #237

### CODE REVIEW

File: app/src/main/java/com/github/steroidteam/todolist/model/todo/Task.java [hunk 1]
```diff
@@ -67,13 +68,15 @@ public Boolean isDone() {
         return done;
     }
 
-    public void setDone(Boolean done) {
+    public Task setDone(Boolean done) {
```
Comment (VPellenc) [inline app/src/main/java/com/github/steroidteam/todolist/model/todo/Task.java:71]: I see that the returned task is used in the ViewModel.
However it looks a bit weird, maybe removeDueDate() could also return the Task for consistency?

**Review by VPellenc — APPROVED** [review]
Hey, I tested the PR and it works well! Here are the steps I used:
- Open the app in my emulator, make some changes, notice there is no latency
- Open the app this time in my phone, observe that all the changes I made on the emulator have indeed been saved
- Go back to the emulator, make more changes but this time without internet connectivity
- Reconnect to the internet
- Go back to the phone, observe that all the changes I made on the emulator have indeed been saved

Only small problem I found is that the Firebase synchronizing (for the changes from the emulator to my phone) could take a bit of time, even though i made sure to restart the app. I don't know if this is a new issue as I never tried this before.

File: app/src/main/java/com/github/steroidteam/todolist/model/todo/Task.java (hunk 1 above)
Comment (theo-damiani) [reply app/src/main/java/com/github/steroidteam/todolist/model/todo/Task.java:71]: Yes, I returned the task itself so it was easier to reuse the new object: like `func(task.setDone(true)).`
But yes it is not consistent with all other methods.

**Comment by theo-damiani** [comment]
> 
> 
> Hey, I tested the PR and it works well! Here are the steps I used:
> 
>     * Open the app in my emulator, make some changes, notice there is no latency
> 
>     * Open the app this time in my phone, observe that all the changes I made on the emulator have indeed been saved
> 
>     * Go back to the emulator, make more changes but this time without internet connectivity
> 
>     * Reconnect to the internet
> 
>     * Go back to the phone, observe that all the changes I made on the emulator have indeed been saved
> 
> 
> Only small problem I found is that the Firebase synchronizing (for the changes from the emulator to my phone) could take a bit of time, even though I made sure to restart the app. I don't know if this is a new issue as I never tried this before.

Thanks for the review!

I think the delay will come from the `syncData()` method in the repository, and as it iterates over all todo and note maybe it can be slow.
And for the changes, keep in mind that the offline functionality is simple:
- all objects in the remote database are downloaded in the local one.
- all objects in the local database that are not in the remote database, are deleted.
- it takes the most recent objects if he is present in both.
So if you go offline and add a new note, it will be not present in the remote database.

**Comment by VPellenc** [comment]
> > Hey, I tested the PR and it works well! Here are the steps I used:
> > ```
> > * Open the app in my emulator, make some changes, notice there is no latency
> > 
> > * Open the app this time in my phone, observe that all the changes I made on the emulator have indeed been saved
> > 
> > * Go back to the emulator, make more changes but this time without internet connectivity
> > 
> > * Reconnect to the internet
> > 
> > * Go back to the phone, observe that all the changes I made on the emulator have indeed been saved
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Only small problem I found is that the Firebase synchronizing (for the changes from the emulator to my phone) could take a bit of time, even though I made sure to restart the app. I don't know if this is a new issue as I never tried this before.
> 
> Thanks for the review!
> 
> I think the delay will come from the `syncData()` method in the repository, and as it iterates over all todo and note maybe it can be slow.
> And for the changes, keep in mind that the offline functionality is simple:
> 
> * all objects in the remote database are downloaded in the local one.
> * all objects in the local database that are not in the remote database, are deleted.
> * it takes the most recent objects if he is present in both.

Okay, good for me!

**Review by shoutizix — APPROVED** [review]
Hey,

Great work !! I've tested this PR on my emulator and on my phone and it worked well, so it's nice !
But as Vincent said there is a lot of delay but I'm not sure we can do so much about it right now so I think this could be merge in master !

**Pull request merged**